src/rsvc/srv_common.c

```diff
@@ -243,8 +243,18 @@ ds_rsvc_get_attr(struct ds_rsvc *svc, struct rdb_tx *tx, rdb_path_t *path,
 				svc->s_name, DP_KEY(&key), rc);
 			goto out_iovs;
 		}
-		iovs[j].iov_buf_len = sizes[i];
-		sizes[i] = iovs[j].iov_len;
+		if (iovs[j].iov_len < sizes[i]) {
+			/* do not xfer more than attr length */
+			bulk_size -= sizes[i] - iovs[j].iov_len;
+			/* return real size sent */
+			sizes[i] = iovs[j].iov_len;
+		} else {
+			/* only return requested size */
+			iovs[j].iov_buf_len = sizes[i];
+			/* only attr size is requested */
+			if (sizes[i] == 0)
+				sizes[i] = iovs[j].iov_len;
+		}
```

Comment:
I see you only want to transfer the real size for each attribute, but the handling here is incorrect.
1. I think you need to change attr_bulk_transfer() - start one crt_bulk_transfer() for each attribute's actual size, and set correct offset/length for each attribute.
2. the sizes setting and bulk_size setting above looks not correct to me.

You may refer obj_bulk_transfer()'s code, just FYI.

Reply:
Hmm, Bruno's change looks right to me. The problem is that the original author fell into the crt pitfall that crt looks at iov_buf_len instead of iov_len when creating a bulk handle, no?

Reply:
in client's attr_bulk_create(), the bulk handle is created by -
2267 >       /* values */                                                            
2268 >       if (sizes != NULL && values != NULL) {                                  
2269 >       >       for (j = 0; j < n; ++j)                                         
2270 >       >       >       if (sizes[j] > 0)                                       
2271 >       >       >       >       d_iov_set(&sgl.sg_iovs[i++],                    
2272 >       >       >       >       >         values[j], sizes[j]);                 
2273 >       }          

i.e. for the attr values, each bulk segment is set to user provided buffer list.

Then at server-side, need to start multi-bulk-transferring to fill to each attr's value's segment, if the real size of some middle segments is less than user provided buffer.
For example, client read 2 attrs, each provide 1KB buffer. But actually at server side, it finds that each attr only with value of 0.5KB. Then should start 2 bulk transfer, 1st transfer to first segment's 0-0.5KB, 2nd transfer to 2nd segment's 0-0.5KB. This is what this patch want to do, no?
or I may mis-understood it...

Reply:
You're right. I mistakenly assumed that the client-side buffer is one continuous segment.

Reply:
Thanks a lot @liuxuezhao and @liw for these very interesting and helpful comments. I think I had to have a look to the similar behaviour in obj_bulk_transfer() before !!... And also it seems to me that attr_bulk_transfer() has been designed to work with a single/common buffer at some point of time...
Will try to push a new commit soon, based on your comments, many thanks again !

Reply:
So you are planning to update this PR @bfaccini ?   I noticed that @liw approved it after that comment.

Reply:
Or perhaps it's just github being weird.  It shows this comment block and afterward, the approval but the approval appears tied to the prior comment from @liw, not the most recent one.

Reply:
> So you are planning to update this PR @bfaccini ?

@jolivier23 , yes I will try to implement the fix for DAOS-6709  based on these comments.

I think @liw had first approved and then felt ok with @liuxuezhao about the way that should be used to fix...